### DISCUSSION TITLE
Auto-generating stdlib/<cloud> libraries

### DISCUSSION THREAD

**Discussion by verdverm** [dagger/dagger#213] [Ideas]
The various cloud libraries are large and ever changing. Given some are making their way into the stdlib, we may want to consider a couple of things:

1. where they live, a) continuing down this path, b) breaking them out
2. to what extent can we auto generate to create and maintain the dagger libraries

- Terraform has manually maintained them (afaik)
- Pulumi looks to generate them from Terraform (I believe they are essentially a frontend for TF + a few other things like docker)

I'm a fan of trying the auto-gen path, but there are gaps in how this would work. We may be able to leverage `cue import go` and TF's Go types for them.

---

Useful references and links:

- https://github.com/GoogleCloudPlatform/terraformer

**Comment by aluzzardi** [comment]
This is a very interesting topic.

Right now we're not impacted by upstream changes: AFAIK the current packages encapsulate CLI access (e.g. `aws cloudformation`, `kubectl`, `helm`), however they do not expose "specifications": kubernetes takes `_` as a spec, it's up to the caller to pass valid kubernetes objects.

The downside is we don't do validation: if you pass an invalid kubernetes config, dagger will gladly accept it and send it to Kubernetes. Eventually `kubectl apply` will fail, only after having executed a bunch of other dagger tasks.

It would be nice to do proper validation of Kubernetes/Cloudformation/etc specifications using Cue since it's something it's really good at.
